### ISSUE TITLE
einsum shape zero

### ISSUE THREAD

**Issue by mariogeiger** [pytorch/pytorch#37628]
## 🐛 Bug

`torch.einsum` crashes when the shape of the input contains zeros. 

## To Reproduce

```python
torch.einsum('ij,j', torch.randn(0, 3), torch.randn(3))
```
```
RuntimeError: shape '[1, 1, 3]' is invalid for input of size 0
```

## Environment

```
PyTorch version: 1.5.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Debian GNU/Linux 10 (buster)
GCC version: (Debian 8.3.0-6) 8.3.0
CMake version: Could not collect

Python version: 3.8
Is CUDA available: No
CUDA runtime version: 9.2.148
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] torch==1.5.0
[pip] torchvision==0.6.0a0+82fd1c8
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.1.243             h6bb024c_0  
[conda] mkl                       2020.0                      166  
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.0.15           py38ha843d7b_0  
[conda] mkl_random                1.1.0            py38h962f231_0  
[conda] numpy                     1.18.1           py38h4f9e942_0  
[conda] numpy-base                1.18.1           py38hde5b4d6_1  
[conda] pytorch                   1.5.0           py3.8_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] torchvision               0.6.0                py38_cu101    pytorch
```

cc @vincentqb @vishwakftw @SsnL @jianyuh

**Comment by ShawnZhong** [comment]
What do you expect, btw? An empty tensor or a better error message?

**Comment by pbelevich** [comment]
We expect numpy's behavior:
```
In [29]: np.einsum('ij,j', np.random.randn(0, 3), np.random.randn(3))
Out[29]: array([], dtype=float64)
```

**Comment by mariogeiger** [comment]
I expect a tensor of shape `[0]`

**Comment by mariogeiger** [comment]
numpy's behavior is not ideal.

If I run `torch.einsum('anything->ijkl', ...)` I expect the output to have dimension 4 and the shape corresponding to the indices `i`, `j`, `k` and `l`.

Example
```
torch.einsum('ijk,k->ij', torch.randn(3, 0, 6), torch.randn(6)) == torch.empty(3, 0)
```

**Comment by mariogeiger** [comment]
This bug is still not solved in 1.7.1

**Comment by heitorschueroff** [comment]
> This bug is still not solved in 1.7.1

@mariogeiger I'm working on a PR that fixes this issue but it had to be reverted after discovering a bug. Expect this to be fixed in 1.8.
